File: csharp/patterns/src/PrototypePatterns/Creational/LazyInitialization/LazyInitializationPattern.cs
namespace PrototypePatterns.Creational.LazyInitialization;

public sealed class LazyDataset
{
    private List<string>? _records;

    public int LoadCount { get; private set; }

    public IReadOnlyList<string> Records
    {
        get
        {
            if (_records is null)
            {
                LoadCount++;
                _records = ["row-1", "row-2", "row-3"];
            }

            return _records;
        }
    }
}

public static class LazyInitializationPattern
{
    public static bool Demo()
    {
        var dataset = new LazyDataset();
        _ = dataset.Records;
        _ = dataset.Records;
        return dataset.LoadCount == 1;
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/creational/lazyinitialization/LazyInitializationPattern.java
package org.prototypepatterns.creational.lazyinitialization;

import java.util.List;

public final class LazyInitializationPattern {
    private LazyInitializationPattern() {
    }

    public static final class LazyDataset {
        private List<String> records;
        private int loadCount;

        public List<String> records() {
            if (records == null) {
                loadCount++;
                records = List.of("row-1", "row-2", "row-3");
            }
            return records;
        }

        public int loadCount() {
            return loadCount;
        }
    }

    public static boolean demo() {
        LazyDataset dataset = new LazyDataset();
        dataset.records();
        dataset.records();
        return dataset.loadCount() == 1;
    }
}
